### ISSUE TITLE
"report_invalid_encoding" error when setting nickname in some cases

### ISSUE THREAD

**Issue by sarsonj** [processone/ejabberd#3587]
## Environment
- ejabberd version: 21.04
- posgresql 12

## Bug description
We are using Posgresql backend with eJabberd. Both client and server encoding is set to UTF-8. Most of UTF strings works, like 99.99% of them. But in some specific cases it fails. I can reproduce it on concrete strings. Here is the example of ejabberd function call, that fails:

```elixir
 :mod_admin_extra.set_nickname("username", "myserver", " ( ͡° ͜ʖ ͡°)")
```

It outputs:

```
** (CaseClauseError) no case clause matching: 
 {:error, {:error, [{:severity, :ERROR}, {86, "ERROR"}, {:code, "22021"},
 {:message, "invalid byte sequence for encoding \"UTF8\": 0xed 0xa1 0xe2"},
 {:file, "wchar.c"}, {:line, 2061}, {:routine, "report_invalid_encoding"}]}}
    (ejabberd 21.4.0) src/mod_admin_extra.erl:1127: :mod_admin_extra.set_nickname/3
```

But when I try to use some other UTF-8 chars, it works without problems:
```elixir
 :mod_admin_extra.set_nickname("username", "myserver", "žluťoučký kůň")
```
:ok

**Comment by prefiks** [comment]
You always use iex shell to enter that? Or do that fail also when using http_api?

Could you show what `:io.format("~p", [YOUR_STRING])` produces? It may just be problem with how iex/console handels those characters.

**Comment by sarsonj** [comment]
No, this happen when I call this through REST API. I just tried to reproduce this using console, so that others can try it as well.

**Comment by badlop** [comment]
I can reproduce that. The command works correctly when using mnesia, but fails when using mysql or pgsql. When it succeeds with mnesia, I checked with a XMPP client that the vcard displays the new nickname correctly.

```
iex(ejabberd@main)1> :mod_admin_extra.set_nickname("admin","localhost", "žluťoučký kůň")      
:ok

iex(ejabberd@main)2> :mod_admin_extra.set_nickname("admin","localhost", " ( ͡° ͜ʖ ͡°)")      
:ok

iex(ejabberd@main)3> :mod_admin_extra.set_nickname("admin","pgsql.localhost", "žluťoučký kůň")
:ok

iex(ejabberd@main)4> :mod_admin_extra.set_nickname("admin","pgsql.localhost", " ( ͡° ͜ʖ ͡°)")    
** (CaseClauseError) no case clause matching:
 {:error, {:error, [{:severity, :ERROR}, {86, "ERROR"}, {:code, "22021"},
 {:message, "invalid byte sequence for encoding \"UTF8\": 0xed 0xa1 0xe2"},
 {:file, "mbutils.c"}, {:line, 1597}, {:routine, "report_invalid_encoding"}]}}
    (ejabberd) src/mod_admin_extra.erl:1126: :mod_admin_extra.set_nickname/3

iex(ejabberd@main)5> :mod_admin_extra.set_nickname("admin","mysql.localhost", "žluťoučký kůň")  
:ok

iex(ejabberd@main)6> :mod_admin_extra.set_nickname("admin","mysql.localhost", " ( ͡° ͜ʖ ͡°)")    
** (CaseClauseError) no case clause matching:
 {:error, {:error, 
    "#HY000Incorrect string value: '\\xED\\xA1\\xE2\\xB0 \\xED...' for column 'lnickname' at row 1"}}
    (ejabberd) src/mod_admin_extra.erl:1126: :mod_admin_extra.set_nickname/3

```

**Comment by prefiks** [comment]
Ok seems that usage of str:to_lower() in mod_vcard:string2lower() is to blame here. It treats string as ascii, which it obviously isn't.

**Comment by prefiks** [comment]
This should be fixed by commit 78c09789f7be66d1844bf8dfd17eb9e7f3851c02
